src/libexpr/value.hh

```diff
@@ -234,6 +274,11 @@ public:
         ExprLambda * fun;
     };
 
+    struct Failed
+    {
+        std::exception_ptr ex;
```

Comment:
I'm currently relying on being able to retry evaluations as more information becomes available, both in hercules-ci-agent and nixops4 (in development).
I see two ways in which we could recover this property.
- Add a "generation" number to the failure and to `EvalState`. Increment the generation number in `EvalState` when more information is known (this is up to the application). When encountering a `tFailed`, retry if the generation nunmber is lower than that in `EvalState`.
- Alternatively, save the progress. For instance an exception thrown in a function call could set the value to `tThunk`, so that recomputing the exception reuses the `Env` and is therefore very cheap. This may perform better when applications retry evaluations, and it shouldn't slow down performance in other cases, because exceptions are almost always rare.
- Ideally, add proper support for suspending the evaluation of one thunk while continuing work on another. This use of exceptions is essentially a workaround for proper suspension, but they're flawed because they don't let us resume the computation efficiently; instead relying on purity. I believe this amounts to
  - Allow switching to a coroutine when missing information is encountered
  - Have a way for the consumer of the evaluator to correlate their `forceValue` call with the coroutine callback, to support reporting
  - Similarly, have a callback when other threads started by the consumer "join" the blocked thread, so those correlations can also be reported by the application

The latter option could lead to the best outcome, but may also be the one that takes most effort.
Progress saving has a similar amount of risk: https://github.com/NixOS/nix/pull/8585 wasn't quite successful, but that might be solvable by saving progress in more parts of the evaluator.
The generation number seems like a simple and reliable quick fix that could unblock this issue.

Reply:
> It'd be safer to put this in a separately allocated struct that we can point to.

That's what it does, see `mkFailed()`.